File: 322. Coin Change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        '''
        dp solution, bottom up
        '''
        dp = [-1] * (amount+1)
        dp[0] = 0
        for i in range(1, amount+1):
            temp = []
            for coin in coins:
                if 0 <= i-coin and dp[i-coin] != -1:
                    temp.append(dp[i-coin]+1)
            dp[i] = min(temp) if temp else -1
        return dp[-1]

File: 322. Coin Change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount+1];
        Arrays.fill(dp, -1);
        dp[0] = 0;
        for (int i=1; i<=amount; i++) {
            int temp = Integer.MAX_VALUE;
            for (int coin : coins) {
                if (i-coin >= 0 && dp[i-coin] >= 0) {
                    temp = Math.min(dp[i-coin], temp);
                }
            }
            if (temp == Integer.MAX_VALUE) {
                dp[i] = -1;
            } else {
                dp[i] = temp+1;
            }
        }
        return dp[amount];
    }
}